File: dehaze-go/internal/repository/message/message_repository.go
package message

import (
	"context"
	"errors"
	"time"

	"github.com/earthyzinc/dehaze-go/internal/model"
	"github.com/earthyzinc/dehaze-go/internal/model/query"
	"gorm.io/gorm"
)

type MessageRepository struct {
	db *gorm.DB
}

func NewMessageRepository(db *gorm.DB) *MessageRepository {
	return &MessageRepository{db: db}
}

func (r *MessageRepository) FindByID(ctx context.Context, id int64) (*model.SysMessage, error) {
	var msg model.SysMessage
	err := r.db.WithContext(ctx).Where("id = ?", id).First(&msg).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &msg, nil
}

func (r *MessageRepository) FindByBizModuleAndBizID(ctx context.Context, bizModule, bizID string) ([]model.SysMessage, error) {
	if bizModule == "" || bizID == "" {
		return nil, nil
	}
	var msgs []model.SysMessage
	err := r.db.WithContext(ctx).
		Where("biz_module = ? AND biz_id = ? AND deleted = 0", bizModule, bizID).
		Find(&msgs).Error
	return msgs, err
}

func (r *MessageRepository) FindPage(ctx context.Context, userID int64, q *query.MessageQuery) ([]model.SysMessage, int64, error) {
	pageNum := q.PageNum
	if pageNum <= 0 {
		pageNum = 1
	}
	pageSize := q.PageSize
	if pageSize <= 0 {
		pageSize = 20
	}

	db := r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("recipient_id = ? AND deleted = 0", userID)

	if q.Type != "" {
		db = db.Where("type = ?", q.Type)
	}
	if q.ReadStatus != nil {
		db = db.Where("read_status = ?", *q.ReadStatus)
	}

	var total int64
	if err := db.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	var msgs []model.SysMessage
	err := db.Order("read_status ASC, create_time DESC").
		Offset((pageNum - 1) * pageSize).Limit(pageSize).
		Find(&msgs).Error
	return msgs, total, err
}

func (r *MessageRepository) SearchPage(ctx context.Context, userID int64, q *query.MessageSearchQuery) ([]model.SysMessage, int64, error) {
	pageNum := q.PageNum
	if pageNum <= 0 {
		pageNum = 1
	}
	pageSize := q.PageSize
	if pageSize <= 0 {
		pageSize = 20
	}

	keyword := "%" + q.Keyword + "%"
	db := r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("recipient_id = ? AND deleted = 0 AND (title LIKE ? OR content LIKE ?)", userID, keyword, keyword)

	var total int64
	if err := db.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	var msgs []model.SysMessage
	err := db.Order("read_status ASC, create_time DESC").
		Offset((pageNum - 1) * pageSize).Limit(pageSize).
		Find(&msgs).Error
	return msgs, total, err
}

func (r *MessageRepository) CountUnread(ctx context.Context, userID int64) (int64, error) {
	var count int64
	err := r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("recipient_id = ? AND read_status = 0 AND deleted = 0", userID).
		Count(&count).Error
	return count, err
}

func (r *MessageRepository) Create(ctx context.Context, msg *model.SysMessage) error {
	return r.db.WithContext(ctx).Create(msg).Error
}

func (r *MessageRepository) CreateBatch(ctx context.Context, msgs []model.SysMessage) ([]int64, error) {
	if len(msgs) == 0 {
		return nil, nil
	}
	if err := r.db.WithContext(ctx).Create(&msgs).Error; err != nil {
		return nil, err
	}
	ids := make([]int64, 0, len(msgs))
	for i := range msgs {
		ids = append(ids, msgs[i].ID)
	}
	return ids, nil
}

func (r *MessageRepository) MarkRead(ctx context.Context, id, userID int64) (int64, error) {
	result := r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("id = ? AND recipient_id = ? AND read_status = 0 AND deleted = 0", id, userID).
		Updates(map[string]interface{}{
			"read_status": 1,
			"read_time":   time.Now(),
			"update_by":   userID,
		})
	return result.RowsAffected, result.Error
}

func (r *MessageRepository) MarkAllRead(ctx context.Context, userID int64, msgType string) (int64, error) {
	db := r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("recipient_id = ? AND read_status = 0 AND deleted = 0", userID)
	if msgType != "" {
		db = db.Where("type = ?", msgType)
	}
	result := db.Updates(map[string]interface{}{
		"read_status": 1,
		"read_time":   time.Now(),
		"update_by":   userID,
	})
	return result.RowsAffected, result.Error
}

func (r *MessageRepository) SoftDelete(ctx context.Context, ids []int64, userID int64) error {
	return r.db.WithContext(ctx).Model(&model.SysMessage{}).
		Where("id IN ? AND recipient_id = ?", ids, userID).
		Update("deleted", 1).Error
}

func (r *MessageRepository) DeleteExpiredBatch(ctx context.Context, before time.Time, batchSize int) (int64, error) {
	var total int64
	for {
		result := r.db.WithContext(ctx).
			Where("expires_at < ?", before).
			Limit(batchSize).
			Delete(&model.SysMessage{})
		if result.Error != nil {
			return total, result.Error
		}
		total += result.RowsAffected
		if result.RowsAffected < int64(batchSize) {
			break
		}
	}
	return total, nil
}

var _ IMessageRepository = (*MessageRepository)(nil)

File: dehaze-python/app/repository/message_repository.py
from datetime import datetime
from typing import Optional

from sqlalchemy import and_, delete, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.entity.sys_message import SysMessage
from app.repository.base import BaseRepository, escape_like


class MessageRepository(BaseRepository[SysMessage]):
    model = SysMessage

    async def get_page(
        self,
        db: AsyncSession,
        recipient_id: int,
        page: int,
        page_size: int,
        type: Optional[str] = None,
        read_status: Optional[int] = None,
    ) -> tuple[list[SysMessage], int]:
        stmt = select(SysMessage).where(
            SysMessage.recipient_id == recipient_id,
            SysMessage.deleted == 0,
        )
        if type:
            stmt = stmt.where(SysMessage.type == type)
        if read_status is not None:
            stmt = stmt.where(SysMessage.read_status == read_status)

        count_stmt = select(func.count()).select_from(stmt.subquery())
        total = (await db.execute(count_stmt)).scalar() or 0

        stmt = stmt.order_by(SysMessage.read_status.asc(), SysMessage.create_time.desc(), SysMessage.id.desc())
        stmt = stmt.offset((page - 1) * page_size).limit(page_size)
        result = await db.execute(stmt)
        items = list(result.scalars().all())
        return items, total

    async def search(
        self,
        db: AsyncSession,
        recipient_id: int,
        keyword: str,
        page: int,
        page_size: int,
    ) -> tuple[list[SysMessage], int]:
        escaped = escape_like(keyword)
        like_pattern = f"%{escaped}%"
        stmt = select(SysMessage).where(
            SysMessage.recipient_id == recipient_id,
            SysMessage.deleted == 0,
            (SysMessage.title.like(like_pattern, escape="\\"))
            | (SysMessage.content.like(like_pattern, escape="\\")),
        )

        count_stmt = select(func.count()).select_from(stmt.subquery())
        total = (await db.execute(count_stmt)).scalar() or 0

        stmt = stmt.order_by(SysMessage.read_status.asc(), SysMessage.create_time.desc(), SysMessage.id.desc())
        stmt = stmt.offset((page - 1) * page_size).limit(page_size)
        result = await db.execute(stmt)
        items = list(result.scalars().all())
        return items, total

    async def count_unread(self, db: AsyncSession, recipient_id: int) -> int:
        stmt = select(func.count()).where(
            SysMessage.recipient_id == recipient_id,
            SysMessage.read_status == 0,
            SysMessage.deleted == 0,
        )
        result = await db.execute(stmt)
        return result.scalar() or 0

    async def get_by_id_and_recipient(
        self,
        db: AsyncSession,
        message_id: int,
        recipient_id: int,
    ) -> Optional[SysMessage]:
        stmt = select(SysMessage).where(
            SysMessage.id == message_id,
            SysMessage.recipient_id == recipient_id,
            SysMessage.deleted == 0,
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def mark_read(self, db: AsyncSession, message_id: int, recipient_id: int) -> bool:
        stmt = (
            update(SysMessage)
            .where(
                SysMessage.id == message_id,
                SysMessage.recipient_id == recipient_id,
                SysMessage.read_status == 0,
                SysMessage.deleted == 0,
            )
            .values(read_status=1, read_time=datetime.now())
        )
        result = await db.execute(stmt)
        return result.rowcount > 0

    async def mark_all_read(
        self,
        db: AsyncSession,
        recipient_id: int,
        type: Optional[str] = None,
    ) -> int:
        stmt = (
            update(SysMessage)
            .where(
                SysMessage.recipient_id == recipient_id,
                SysMessage.read_status == 0,
                SysMessage.deleted == 0,
            )
            .values(read_status=1, read_time=datetime.now())
        )
        if type:
            stmt = stmt.where(SysMessage.type == type)
        result = await db.execute(stmt)
        return result.rowcount

    async def soft_delete_by_ids_and_recipient(
        self,
        db: AsyncSession,
        ids: list[int],
        recipient_id: int,
    ) -> int:
        stmt = (
            update(SysMessage)
            .where(
                SysMessage.id.in_(ids),
                SysMessage.recipient_id == recipient_id,
            )
            .values(deleted=1)
        )
        result = await db.execute(stmt)
        return result.rowcount

    async def find_by_biz(
        self,
        db: AsyncSession,
        biz_module: str,
        biz_id: str,
    ) -> list[SysMessage]:
        stmt = select(SysMessage).where(
            SysMessage.biz_module == biz_module,
            SysMessage.biz_id == biz_id,
            SysMessage.deleted == 0,
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def batch_create(self, db: AsyncSession, messages: list[SysMessage]) -> list[SysMessage]:
        db.add_all(messages)
        await db.flush()
        return messages

    async def delete_expired(self, db: AsyncSession, now: datetime, batch_size: int = 500) -> int:
        total_deleted = 0
        while True:
            id_stmt = select(SysMessage.id).where(
                SysMessage.expires_at < now,
            ).limit(batch_size)
            id_result = await db.execute(id_stmt)
            ids = [row[0] for row in id_result.fetchall()]
            if not ids:
                break
            stmt = delete(SysMessage).where(SysMessage.id.in_(ids))
            result = await db.execute(stmt)
            total_deleted += result.rowcount
            if len(ids) < batch_size:
                break
        return total_deleted


message_repository = MessageRepository()
